File: packages/client/src/schemas/media.ts
// Runtime schemas for the media / catalogue domain mirrors of the ts-rs
// generated wire types (see `../generated`, the single source of truth).
//
// Same conventions as `./accounts.ts`: each schema mirrors a Rust struct, adds
// runtime validation + branded ids, and carries a compile-time drift guard.
// `true satisfies SameKeys<…>` breaks the build if a field is added/removed in
// Rust and `gen:types` is rerun without updating the schema. Tagged unions use
// `z.discriminatedUnion` and skip the drift line (it yields false failures).
//
// Schemas are defined in dependency order: a type used as a nested field appears
// above the schema that references it.

import { z } from 'zod';
import { ItemId, LibraryId, ShowId } from './ids';

/** What sort of thing a media item is. */
export const MediaKind = z.enum(['movie', 'episode', 'video']);

/** What a [`Marker`] segment is (serialized lowercase). */
export const MarkerKind = z.enum(['intro', 'credits']);

/** Library classification, derived from the kinds of items it holds. */
export const LibraryKind = z.enum(['movies', 'shows', 'mixed']);

/** UI grouping bucket for a job (serializes lowercase). */
export const Category = z.enum([
  'maintenance',
  'library',
  'recommendations',
  'pipeline',
  'acquisition',
]);

/** Video stream description (fields may be null when unknown). */
export const VideoTrack = z.object({
  codec: z.string(),
  width: z.number().nullable(),
  height: z.number().nullable(),
  hdr: z.boolean(),
  bitDepth: z.number().nullable(),
});
export type VideoTrack = z.infer<typeof VideoTrack>;

/** One audio stream/track. */
export const AudioTrack = z.object({
  index: z.number(),
  codec: z.string(),
  channels: z.number().nullable(),
  language: z.string().nullable(),
  title: z.string().nullish(),
  default: z.boolean(),
});
export type AudioTrack = z.infer<typeof AudioTrack>;

/** A subtitle track. */
export const SubtitleTrack = z.object({
  language: z.string().nullable(),
  codec: z.string(),
});
export type SubtitleTrack = z.infer<typeof SubtitleTrack>;

/** One physical file backing a logical [`MediaItem`]. `id` is a `short_hash` of
 * the absolute path (opaque, per-file) not a media-item id. */
export const MediaFile = z.object({
  id: z.string(),
  relPath: z.string().nullable(),
  container: z.string(),
  durationMs: z.number().nullable(),
  video: VideoTrack.nullable(),
  audio: AudioTrack.nullable(),
  audioTracks: z.array(AudioTrack),
  subtitles: z.array(SubtitleTrack),
  size: z.number().nullable(),
  edition: z.string().nullish(),
  probed: z.boolean(),
});
export type MediaFile = z.infer<typeof MediaFile>;

/** One timed segment of an episode (intro / credits), in milliseconds. */
export const Marker = z.object({
  kind: MarkerKind,
  startMs: z.number(),
  endMs: z.number(),
});
export type Marker = z.infer<typeof Marker>;

/** One top-billed cast member. */
export const CastMember = z.object({
  name: z.string(),
  character: z.string().nullable(),
  profileUrl: z.string().nullish(),
});
export type CastMember = z.infer<typeof CastMember>;

/** One key crew member (director, writer, creator). */
export const CrewMember = z.object({
  name: z.string(),
  job: z.string(),
  profileUrl: z.string().nullish(),
});
export type CrewMember = z.infer<typeof CrewMember>;

/** Resolved provider metadata for one movie or show. `imdbId` is an opaque
 * external id (plain string); `tmdbId` is numeric. */
export const Metadata = z.object({
  provider: z.string(),
  tmdbId: z.number(),
  imdbId: z.string().nullable(),
  title: z.string().nullable(),
  tagline: z.string().nullable(),
  overview: z.string().nullable(),
  releaseDate: z.string().nullable(),
  genres: z.array(z.string()),
  rating: z.number().nullable(),
  posterUrl: z.string().nullable(),
  backdropUrl: z.string().nullable(),
  logoUrl: z.string().nullable(),
  themeUrl: z.string().nullish(),
  cast: z.array(CastMember).nullish(),
  crew: z.array(CrewMember).nullish(),
  tmdbUrl: z.string(),
});
export type Metadata = z.infer<typeof Metadata>;

/** A single playable media item. `defaultFileId` is an opaque file id (plain
 * string); `id`/`showId` are branded. */
export const MediaItem = z.object({
  id: ItemId,
  title: z.string(),
  kind: MediaKind,
  year: z.number().nullable(),
  durationMs: z.number().nullable(),
  container: z.string(),
  video: VideoTrack.nullable(),
  audio: AudioTrack.nullable(),
  audioTracks: z.array(AudioTrack),
  subtitles: z.array(SubtitleTrack),
  library: z.string(),
  showId: ShowId.nullable(),
  showTitle: z.string().nullable(),
  season: z.number().nullable(),
  episode: z.number().nullable(),
  episodeEnd: z.number().nullable(),
  episodeTitle: z.string().nullable(),
  relPath: z.string().nullable(),
  addedAt: z.string(),
  metadata: Metadata.nullish(),
  files: z.array(MediaFile),
  defaultFileId: z.string().nullish(),
  markers: z.array(Marker).nullish(),
});
export type MediaItem = z.infer<typeof MediaItem>;

/** One season's worth of episodes, sorted by episode number. */
export const Season = z.object({
  number: z.number(),
  episodes: z.array(MediaItem),
  cast: z.array(CastMember).nullish(),
});
export type Season = z.infer<typeof Season>;

/** A TV show aggregate (not a file), built by grouping episodes during a scan. */
export const Show = z.object({
  id: ShowId,
  title: z.string(),
  year: z.number().nullable(),
  library: z.string(),
  seasonCount: z.number(),
  episodeCount: z.number(),
  video: VideoTrack.nullable(),
  addedAt: z.string(),
  metadata: Metadata.nullish(),
  progress: z.number().nullish(),
});
export type Show = z.infer<typeof Show>;

/** `GET /api/shows/:id` payload: a show plus its seasons. */
export const ShowDetail = z.object({
  show: Show,
  seasons: z.array(Season),
});
export type ShowDetail = z.infer<typeof ShowDetail>;

/** A scanned library root. `id` is an opaque library id (plain string). */
export const Library = z.object({
  id: LibraryId,
  name: z.string(),
  kind: LibraryKind,
  path: z.string(),
  itemCount: z.number(),
});
export type Library = z.infer<typeof Library>;

/** One row of a user's playback progress. */
export const ProgressEntry = z.object({
  itemId: ItemId,
  positionMs: z.number(),
  durationMs: z.number().nullable(),
  updatedAt: z.string(),
});
export type ProgressEntry = z.infer<typeof ProgressEntry>;

/** A "continue watching" entry: the resumable item plus where to resume from. */
export const ContinueItem = z.object({
  item: MediaItem,
  positionMs: z.number(),
  durationMs: z.number().nullable(),
  updatedAt: z.string(),
});
export type ContinueItem = z.infer<typeof ContinueItem>;

/** The episode to play to continue a show (`GET /api/shows/:id/up-next`). */
export const UpNext = z.object({
  item: MediaItem,
  resume: z.boolean(),
});
export type UpNext = z.infer<typeof UpNext>;

/** One rail entry: a movie/video (a [`MediaItem`]) or a whole show (a [`Show`]),
 * distinguished by the `type` tag. */
// drift: runtime-checked (tagged union)
export const SectionItem = z.discriminatedUnion('type', [
  z.object({ type: z.literal('movie'), item: MediaItem }),
  z.object({ type: z.literal('show'), show: Show }),
]);
export type SectionItem = z.infer<typeof SectionItem>;

/** One catalogue rail: a titled row of [`SectionItem`]s. */
export const Section = z.object({
  id: z.string(),
  title: z.string(),
  reason: z.string().nullable(),
  items: z.array(SectionItem),
});
export type Section = z.infer<typeof Section>;

/** The status of one treatment (stage) applied to a single catalog element. */
export const Treatment = z.object({
  key: z.string(),
  status: z.string(),
  error: z.string().nullable(),
});
export type Treatment = z.infer<typeof Treatment>;

/** `GET /api/health`. */
export const Health = z.object({
  status: z.string(),
  version: z.string(),
  ffprobe: z.boolean(),
  libraries: z.number(),
  items: z.number(),
  shows: z.number(),
});
export type Health = z.infer<typeof Health>;

/** Server identity + uptime for the admin sidebar status card. */
export const ServerInfo = z.object({
  name: z.string(),
  hostname: z.string(),
  version: z.string(),
  uptimeSec: z.number(),
  online: z.boolean(),
  sessions: z.number(),
});
export type ServerInfo = z.infer<typeof ServerInfo>;
export type Category = z.infer<typeof Category>;
export type LibraryKind = z.infer<typeof LibraryKind>;
export type MarkerKind = z.infer<typeof MarkerKind>;
export type MediaKind = z.infer<typeof MediaKind>;

File: server/src/domain/media.rs
//! Media catalog types: kinds, stream descriptions, files, items and shows.
//!
//! The JSON shape here is a public contract web/TV clients depend on it, so
//! field names and casing must not drift.

use serde::{Deserialize, Serialize};

use crate::domain::metadata::{CastMember, Metadata};

/// What sort of thing a media item is.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Kind {
    Movie,
    Episode,
    Video,
}

/// Video stream description (best-effort; fields may be null when unknown).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoStream {
    pub codec: String,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub hdr: bool,
    #[serde(rename = "bitDepth")]
    pub bit_depth: Option<u32>,
}

/// One audio stream/track. An item can carry several (e.g. EN + FR, or a
/// director's commentary); `index` is the **audio-relative** position (0-based
/// among audio streams only), which is exactly what ffmpeg's `-map 0:a:<index>`
/// selector expects when remuxing a chosen track.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioStream {
    /// Audio-relative stream index (0 = first audio track). Drives track
    /// selection (`-map 0:a:<index>`) on the server's per-track HLS remux.
    #[serde(default)]
    pub index: u32,
    pub codec: String,
    pub channels: Option<u32>,
    pub language: Option<String>,
    /// Human label from the stream's `title` tag ("Commentary", "Director's
    /// Cut", …), when present.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Whether the container marks this as the default audio track.
    #[serde(default)]
    pub default: bool,
}

/// A subtitle track.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubtitleTrack {
    pub language: Option<String>,
    pub codec: String,
}

/// One physical file backing a logical [`MediaItem`]. A single item can have
/// several of these (Director's Cut + Theatrical, 1080p + 4K, …); they all share
/// the same logical item id but each maps to a distinct file on disk.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaFile {
    /// `short_hash(abs_path)` stable per physical file.
    pub id: String,
    #[serde(rename = "relPath")]
    pub rel_path: Option<String>,
    pub container: String,
    #[serde(rename = "durationMs")]
    pub duration_ms: Option<u64>,
    pub video: Option<VideoStream>,
    /// Representative (first/default) audio track kept for codec badges and
    /// backward compatibility with clients that read `audio.codec` directly.
    pub audio: Option<AudioStream>,
    /// Every audio track on this file, in container order. Drives the player's
    /// audio-track picker. Empty when unprobed or for pre-`audioTracks` rows.
    #[serde(rename = "audioTracks", default)]
    pub audio_tracks: Vec<AudioStream>,
    pub subtitles: Vec<SubtitleTrack>,
    pub size: Option<u64>,
    /// Best-effort label parsed from the filename ("Director's Cut", "Extended",
    /// "Remux", "4K", "1080p", …). `None` when nothing notable is detected.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub edition: Option<String>,
    /// Whether ffprobe has run on this file yet (phase 2). Until then the stream
    /// fields above are null.
    pub probed: bool,
    /// Absolute path on disk. Internal only never serialized to clients.
    #[serde(skip)]
    pub abs_path: Option<String>,
}

/// A single playable media item.
///
/// `rel_path` is relative to the owning media directory. Demo/seed items have
/// `rel_path == None` and cannot be streamed.
///
/// An item can be backed by multiple physical [`MediaFile`]s. The top-level
/// `video`/`audio`/`duration_ms`/`container`/`subtitles`/`abs_path` fields mirror
/// the **representative file** (the highest-resolution probed file) for backward
/// compatibility with clients that read `item.video.codec` directly.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaItem {
    pub id: String,
    pub title: String,
    pub kind: Kind,
    pub year: Option<u32>,
    #[serde(rename = "durationMs")]
    pub duration_ms: Option<u64>,
    pub container: String,
    pub video: Option<VideoStream>,
    /// Representative (first/default) audio track kept for badges and
    /// backward compatibility. The full list is `audio_tracks`.
    pub audio: Option<AudioStream>,
    /// Every audio track of the representative file, for the audio-track picker.
    #[serde(rename = "audioTracks", default)]
    pub audio_tracks: Vec<AudioStream>,
    pub subtitles: Vec<SubtitleTrack>,
    pub library: String,
    // --- show / episode grouping (null for movies) ---
    #[serde(rename = "showId")]
    pub show_id: Option<String>,
    #[serde(rename = "showTitle")]
    pub show_title: Option<String>,
    pub season: Option<u32>,
    pub episode: Option<u32>,
    /// Last episode number for multi-episode files (`S01E02-E03`).
    #[serde(rename = "episodeEnd")]
    pub episode_end: Option<u32>,
    #[serde(rename = "episodeTitle")]
    pub episode_title: Option<String>,
    #[serde(rename = "relPath")]
    pub rel_path: Option<String>,
    #[serde(rename = "addedAt")]
    pub added_at: String,
    /// TMDB catalog metadata (poster/backdrop/overview/IDs). `None` until the
    /// background enrichment pass resolves it. Movies only; episodes inherit
    /// their show's metadata.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    /// Absolute path on disk. Internal only never serialized to clients.
    /// Mirrors the default/representative file's path so `/stream` keeps working.
    #[serde(skip)]
    pub abs_path: Option<String>,
    /// Every physical file backing this logical item.
    #[serde(default)]
    pub files: Vec<MediaFile>,
    /// Id of the representative ("default") file the one `/stream` serves and
    /// whose stream info populates the top-level fields. `None` until at least
    /// one file exists.
    #[serde(rename = "defaultFileId", default, skip_serializing_if = "Option::is_none")]
    pub default_file_id: Option<String>,
    /// Intro / credits segment markers (episodes only). Drives the "skip intro"
    /// button and the credits-triggered "next episode" card. Empty until resolved
    /// from chapters or the audio-fingerprint job.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub markers: Vec<Marker>,
}

/// What a [`Marker`] segment is. Serialized lowercase (`"intro"` / `"credits"`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MarkerKind {
    Intro,
    Credits,
}

/// One timed segment of an episode (intro / credits), in milliseconds.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Marker {
    pub kind: MarkerKind,
    #[serde(rename = "startMs")]
    pub start_ms: u64,
    #[serde(rename = "endMs")]
    pub end_ms: u64,
}

/// A TV show aggregate (not a file). Built by grouping episodes during a scan.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Show {
    pub id: String,
    pub title: String,
    pub year: Option<u32>,
    pub library: String,
    #[serde(rename = "seasonCount")]
    pub season_count: u32,
    #[serde(rename = "episodeCount")]
    pub episode_count: u32,
    /// Representative video info (from an episode) for quality badges.
    pub video: Option<VideoStream>,
    #[serde(rename = "addedAt")]
    pub added_at: String,
    /// TMDB catalog metadata (poster/backdrop/overview/IDs). `None` until the
    /// background enrichment pass resolves it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    /// Per-user series-completion percent (0–100), filled by the catalogue
    /// endpoints when the request is authenticated. `None` for anonymous requests
    /// or shows with no progress drives the progress bar on show cards.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub progress: Option<u8>,
}

/// One season's worth of episodes, sorted by episode number.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Season {
    pub number: u32,
    pub episodes: Vec<MediaItem>,
    /// Season-specific cast (TMDB season credits), resolved during enrichment.
    /// Empty until enriched or when the provider returned none.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub cast: Vec<CastMember>,
}

/// `GET /api/shows/:id` payload: a show plus its seasons.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShowDetail {
    pub show: Show,
    pub seasons: Vec<Season>,
}
